datafusion-cli/tests/cli_integration.rs

```diff
@@ -17,42 +17,223 @@
+    let access_key_id =
+        env::var("AWS_ACCESS_KEY_ID").expect("AWS_ACCESS_KEY_ID is not set");
+    let secret_access_key =
+        env::var("AWS_SECRET_ACCESS_KEY").expect("AWS_SECRET_ACCESS_KEY is not set");
+
+    let endpoint_url = env::var("AWS_ENDPOINT").expect("AWS_ENDPOINT is not set");
+    let region = Region::new(env::var("AWS_REGION").unwrap_or("df-test".to_string()));
+
+    let allow_non_test_credentials = env::var("ALLOW_NON_TEST_CREDENTIALS")
+        .map(|v| v == "1")
+        .unwrap_or(false);
+
+    if allow_non_test_credentials
+        || !access_key_id.starts_with("TEST-")
+        || !secret_access_key.starts_with("TEST-")
+    {
+        panic!("Refusing with non-test credentials. Either set ALLOW_NON_TEST_CREDENTIALS=1 or add TEST- for AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY");
+    }
+
+    let creds = Credentials::new(access_key_id, secret_access_key, None, None, "test");
+    let config = aws_sdk_s3::Config::builder()
+        .credentials_provider(creds)
+        .endpoint_url(endpoint_url)
+        .region(region)
+        .behavior_version(BehaviorVersion::v2024_03_28())
+        .build();
+
+    aws_sdk_s3::Client::from_conf(config)
+}
+
+async fn create_bucket(bucket_name: &str, client: &aws_sdk_s3::Client) {
+    match client.head_bucket().bucket(bucket_name).send().await {
+        Ok(_) => {}
+        Err(SdkError::ServiceError(err))
+            if matches!(
+                err.err(),
+                aws_sdk_s3::operation::head_bucket::HeadBucketError::NotFound(_)
+            ) =>
+        {
+            client
+                .create_bucket()
+                .bucket(bucket_name)
+                .send()
+                .await
+                .expect("Failed to create bucket");
+        }
+        Err(e) => panic!("Failed to head bucket: {:?}", e),
+    }
+}
+
+async fn move_file_to_bucket(
+    from_path: &str,
+    to_path: &str,
+    bucket_name: &str,
+    client: &aws_sdk_s3::Client,
+) {
+    let body =
+        aws_sdk_s3::primitives::ByteStream::from_path(std::path::Path::new(from_path))
+            .await
+            .expect("Failed to read file");
+
+    client
+        .put_object()
+        .bucket(bucket_name)
+        .key(to_path)
+        .body(body)
+        .send()
+        .await
+        .expect("Failed to put object");
+}
+
+#[tokio::test]
+async fn test_cli() {
+    if env::var("TEST_STORAGE_INTEGRATION").is_err() {
+        eprintln!("Skipping external storages integration tests");
+        return;
+    }
+
+    let client = s3_client().await;
+    create_bucket("cli", &client).await;
+    move_file_to_bucket(
+        "../datafusion/core/tests/data/cars.csv",
+        "cars.csv",
+        "cli",
+        &client,
+    )
+    .await;
+
+    let settings = make_settings();
+    let _bound = settings.bind_to_scope();
+
+    glob!("sql/*.sql", |path| {
+        let input = fs::read_to_string(path).unwrap();
+        assert_cmd_snapshot!(cli().pass_stdin(input))
```

Comment:
It is absolutely possible! My idea was to make it closer to `slt` and separate the code from the data. Setting up the suite is quite hard (we need to create a new user, upload a test file, etc.), but with the current approach, we only need to do it once rather than for every single test.

However, I'm happy to make changes if you'd prefer it to be inline?

Reply:
I personally prefer it to be inline, but perhaps we can do that as a follow on PR 

(the way we do this in influxdb_iox is that we have a test function that returns a string and then compare the string with the `assert_yaml_snapshot` -- that way the setup code isn't replicated and we can still have the tests inline